sdk/core/azure-core/CHANGELOG.md

```diff
@@ -10,6 +10,7 @@
 - Renamed `NoRevoke` to `EnableCertificateRevocationListCheck` for `Azure::Core::Http::CurlTransportSSLOptions`.
 - Renamed `GetString()` to `ToString()` in `Azure::Core::DateTime`.
 - Renamed `GetUuidString()` tp `ToString()` in `Azure::Core::Uuid`.
+- Moved `Url` to the internals. Not for public access.
```

Comment:
We should talk about this. I think it is premature to make internal a fundamental type like `Url` which is important for HTTP request/responses.

What benefit do we get from making it internal?

If we want to expose this type as an exchange type that public client libraries accept as a parameter in the future, what would be our plan for this type then? Would we move it to be public or create a duplicate type in a different namespace?

Reply:
> a fundamental type like Url which is important for HTTP request/responses.

Why do you think it is fundamental?
For an end-user?  It is not!  The SDK client abstract the Http request as policies that runs sequentially to complete a REST operation. This is the fundamentals of the SDK.  It only exposes options for the policies.
So, the way that the SDK client works internally to complete this task is an implementation detail.
That's why the Pipepline is not exposed, and the URL is under that category as well. 
Ideally, the Request should be also internal, but it is in the signature of an Http policy. Supporting the extensible-policies features to allow end-users to add custom policies makes us let a door open for end-users to our implementation details for how we define an http request.

> What benefit do we get from making it internal?

Document,  test and support with the same criteria as we do for public APIs.
Adjust the implementation details and the API for the URL to be the best experience for Service client developers (storage, keyvault, etc) without considering end-users experience/usability.

 > what would be our plan for this type then?

This question is more generalized, and not only for this specific case. Will we allow that?  Because the same thing can be asked to any internal component, like the pipeline or anything. Basically, will we ever allow to expose implementation details to the Public? 
Just think about the Url as a shared implementation detail for all SDK clients.

Reply:
Ask the Java team why they decided to make it public. 

You are over-indexing on the "document, test, support" benefit (which arguably isn't that high to begin with and comes with its own trade-offs), and focusing on the state of the SDK today, rather than creating core types that would be needed for many more client libraries to come.

Look at all the language SDKs (can start with Java/.NET) and see how many public APIs accept `Uri`/`Url` and related types. In many cases, it is treated as an exchange type, instead of string. That's why such a type is fundamental to the SDK.
For example:
https://github.com/Azure/azure-sdk-for-net/blob/2b594335737a65cb8f36543ef817deaba5c43f4b/sdk/keyvault/Microsoft.Azure.KeyVault/src/Customized/Authentication/HttpBearerChallengeCache.cs#L77

https://github.com/Azure/azure-sdk-for-net/blob/2b594335737a65cb8f36543ef817deaba5c43f4b/sdk/storage/Azure.Storage.Blobs/src/BlobUriBuilder.cs#L172

https://github.com/Azure/azure-sdk-for-java/blob/5672b860b15aa32d1dd5b559f47c3cb1058f0166/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/Uri.java#L9

https://github.com/Azure/azure-sdk-for-java/blob/5672b860b15aa32d1dd5b559f47c3cb1058f0166/sdk/mediaservices/microsoft-azure-media/src/main/java/com/microsoft/windowsazure/services/media/implementation/ResourceLocationManager.java#L38

Reply:
Is the one you shared from Java, only for cosmos? (and the other one is track1), It's not part of Core.

Think about the end-user mostly.  Consider these scenarios:
1.- As an end-user, I want to run an SDK client method (or even create a new sdk client) and I am asked to provide a `Azure::Url`.  It means that before learning anything about the sdk client, I need to learn about the Azure::Url.  Basically, the Url is becoming as relevant and significant as the credentials!
And this is not like Java or .Net where I don't care about lifetime. For C++, I need to learn if the Url will create things in the heap that I need to eventually free.  I need to know how expensive it to pass Url as parameter. 
I need to learn if I need to keep it alive or if a copy will be created. Etc etc..   All the things that developers already know about std::string.

2- ok, I have learned about the Url and I am ready to call an API operation. I have created my client and I called client.Foo().
The return type is `<T>` and it has a Url field.  I am interested in using the Url field somewhere in my app. What are my options now?  Do I need to carry that Url type all around my code to avoid making an std::string copy to get the value I need out of the Url?   Do I need to keep the Url someplace alive where I can just get a std::string const& to its value?  why can't I just get an std::string which I can std::move() where ever I need?

3.- Ok, probably I have found one benefit. One that I have received an `Url` object, I can use it to create a new sdk client!  or call any client method which takes an Url !   But wait?  I don't want to pass it as value and make copies, that's too slow! I want to move-construct!  But if I do that, I will lost the Url.  So, it was looking as a cool feature at the beginning, but it is actually an inconvenient mechanism!.  I will just call ToString() from the Url and use it to construct a new Url instead.

Reply:
The lifetime of objects applies to many types in C++ and the SDK, I am not sure how `Url` is unique in that regard at all. I could make the same arguments for DateTime and Etag, for example.

I shared a couple of links, and recommend you find some for yourself, rather than picking apart the specifics of 1-2 that don't fit. What about the other examples?

Are we saying we never want any future SDK to accept a `Url` type in a public API for any service, even when it made sense in other languages? Why would we make that decision for them, forever? Azure::Core is meant to provide common types used by multiple clients, `Url` is one such type.

Reply:
> The lifetime of objects applies to many types in C++

Yes, that's why the less types you add for the public, the less you need to care about all the docs and lifetime of things.

> I could make the same arguments for DateTime and Etag, for example.

Whe main difference is that Url would become as important as a Credential type. An entry point for the SDK client.  You need it to create a new client.  We would be giving it more importance than the API methods 🤣 

>  recommend you find some for yourself

I am convinced about exposing Url in C++ is a very bad idea. No matter what other languages have done or will do.  For the benefit of the C++ customers, I would ask for language disagreement here (which is allowed when the language requires it).

> Are we saying we never want any future SDK to accept a Url type in a public API for any service

Yes, just how we have update the Core to block services for creating the pipeline directly, the same way I am expecting Core to keep the Url as an implementation detail shared across services clients only. Never public

Reply:
I will do some more investigation/reflection and explore prior art, and get back to you.

If we want to make it internal, please lets rename it and reserve the good `Url` name for the public type we might add in the future. Maybe `UrlUtility`?